### ISSUE TITLE
evidence of a possible memory issue within the Snakemake build

### ISSUE THREAD

**Issue by kvarforl** [RTXteam/RTX#1117]
during KG2.3.5 build, the rule Stats was randomly interrupted by a bash kill signal. On the first restart, the same thing happened, but on the second restart, everything was fine.

this is the relevant portion of `build-kg2-snakemake.log`:
```
 79 [Fri Oct 23 03:30:28 2020]
 80 rule Stats:
 81     input: /home/ubuntu/kg2-build/kg2.json
 82     output: /home/ubuntu/kg2-build/kg2-report.json
 83     jobid: 2
 84 
 85 /bin/bash: line 1:  6245 Killed                  /home/ubuntu/kg2-venv/bin/python3 -u /home/ubuntu/kg2-code/report    _stats_on_json_kg.py /home/ubuntu/kg2-build/kg2.json /home/ubuntu/kg2-build/kg2-report.json
 86 [Fri Oct 23 03:49:43 2020]
 87 Error in rule Stats:
 88     jobid: 2
 89     output: /home/ubuntu/kg2-build/kg2-report.json
 90     shell:
 91         /home/ubuntu/kg2-venv/bin/python3 -u /home/ubuntu/kg2-code/report_stats_on_json_kg.py /home/ubuntu/kg2-bui    ld/kg2.json /home/ubuntu/kg2-build/kg2-report.json
 92         (exited with non-zero exit code)
 93 
 94 Shutting down, this might take some time.
 95 Exiting because a job execution failed. Look above for error message
 96 Complete log: /home/ubuntu/.snakemake/log/2020-10-23T033028.362286.snakemake.log
```

**Comment by saramsey** [comment]
I wonder if this is the Linux kernel "OOM Killer" at work.
https://www.kernel.org/doc/gorman/html/understand/understand016.html

**Comment by saramsey** [comment]
Has this issue recurred, since last October?  Did it come up in the 2.5.0 or 2.5.1 builds?  I note that 2.3.5 was done on `kg2lindsey` which is the same instance t ype as was used for the 2.5.0 and 2.5.1 builds. So if the bug hasn't happened again, I suppose it is a hard-to-reproduce timing-dependent bug with the parallel build. By "bug", I mean that I am conjecturing that the Snakemake parallel build is permitting two steps of the build to happen at once that could exhaust the system memory.

**Comment by kvarforl** [comment]
I agree! I haven't seen this since October, so I think it would be fine to close out!

**Comment by kvarforl** [comment]
Noting that this happened again with KG2.6.0, but in the Simplify_Nodes rule:
```
 50 [Tue Apr 20 20:15:37 2021]
 51 rule Simplify_Nodes:
 52     input: /home/ubuntu/kg2-build/kg2-simplified.json
 53     output: /home/ubuntu/kg2-build/kg2-simplified-nodes.json
 54     jobid: 6
 55
 56 /bin/bash: line 1: 24363 Killed                  /home/ubuntu/kg2-venv/bin/python3 -u /home/ubuntu/kg2-code/get_nodes_json_from_kg_json.py /home/ubuntu/kg2-build    /kg2-simplified.json /home/ubuntu/kg2-build/kg2-simplified-nodes.json
 57 [Tue Apr 20 20:38:48 2021]
```

Restarting to see if it repeats :)

**Comment by kvarforl** [comment]
hmmm, it is unfortunately recurring

**Comment by saramsey** [comment]
This is very likely the Linux kernel OOM process killer.
https://docs.memset.com/other/linux-s-oom-process-killer

Did restarting allow the build to complete?

**Comment by kvarforl** [comment]
yep, I'd believe it! It seems strange that it's failing in `get_nodes_json_from_kg_json.py` but didn't fail when running `filter_and_remap_predicates.py`. Does anything get added from one step to the next? I guess I'm confused at how it's suddenly too large

**Comment by kvarforl** [comment]
Some file sizes:
Filename | KG2.5.2 | KG2.6.0 |
|---|---|---|
kg2.json | 49 GB | 59 GB|
kg2-simplified.json | 49GB | 58 GB |

**Comment by kvarforl** [comment]
okay it made it through the area that was failing earlier; htop said it used 242 of the 250 available memory, so I understand how it was on the edge.

I'm going to increase the jensenlab zscore cuttoff to 3 anyway, until we see how popular the edges are. if they're in high demand, we can add them back in :)
